### PR TITLE
feat: add machine configurator

### PR DESCRIPTION
**Pull request by rumeshwick**
**Changes** 
Added Machine Ids configurator with supporting saving the configuration in firebase and localstorage.
Added validation not open widget customizer if there are no configured machines.
Added rule to redirect user to settings page when user goes to overview while there are no configured machines

### CODE REVIEW

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    defaultValues,
+  });
+
+  /**
+   * This will open the configurator popup on machine settings icon click.
+   * restoreMachineConfig function is used to fill the form fields with previously saved values
+   */
+  const handleClickOpen = () => {
+    setOpen(true);
+    setTimeout(() => {
```
Comment (man-csl) [inline src/components/MachineSelector/Configurator/MachineConfigurator.tsx]: Whats the purpose of delaying this?

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    const savedMachineConfigStr = await getMachineIds(onError);
+    if (savedMachineConfigStr) {
+      setAssetExternalIds(savedMachineConfigStr.split(','));
+    }
+  };
+
+  /**
+   * This function is used to validate machine saving.
+   */
+  const isValidSaving = () => {
```
Comment (man-csl) [inline src/components/MachineSelector/Configurator/MachineConfigurator.tsx]: Both comment and name of the function is not very clear.

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+   */
+  const saveMachineConfig = async () => {
+    if (isValidSaving()) {
+      const actionStatus = await saveMachineIds(
+        assetExternalIds.join(','),
+        onError
+      );
+      if (actionStatus) {
+        handleClose();
+        props.updateAssets();
```
Comment (man-csl) [inline src/components/MachineSelector/Configurator/MachineConfigurator.tsx:162]: What is the significant of this?

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+  /**
+   * This method is used to add asset id to the list
+   */
+  const addAssetId = (assetId: string) => {
+    setAssetExternalIds([assetId, ...assetExternalIds]);
+    reset();
+  };
+
+  /**
+   * This method is used to validate asset id before it adding to the list
```
Comment (man-csl) [inline src/components/MachineSelector/Configurator/MachineConfigurator.tsx]: it adding to -> it is added to

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  };
+
+  /**
+   * This function fires on form submit
+   */
+  const onSubmit = (data: FormField) => {
+    validateAssetId(data.assetExternalId);
+  };
+
+  return (
```
Comment (man-csl) [inline src/components/MachineSelector/Configurator/MachineConfigurator.tsx:215]: Some comments at-least describing different parts would be helpful.

File: src/components/widgets/timeSeries/TSFancyNumeric/TSFancyNumeric.test.tsx [hunk 6]
```diff
@@ -5,20 +5,22 @@ import { TSFancyNumericMockProps } from 'mocks/widgetsMockData/circularProgressM
...
 describe('CircularProgress', () => {
-  it('should render progress precentage correctly', () => {
-    const { getByText } = render(
-      <TSFancyNumeric {...TSFancyNumericMockProps[0]} />
-    );
-    expect(
-      getByText(`${TSFancyNumericMockProps[0].precentage.toString()}%`)
-    ).toBeInTheDocument();
-  });
+  // it('should render progress precentage correctly', () => {
```
Comment (man-csl) [inline src/components/widgets/timeSeries/TSFancyNumeric/TSFancyNumeric.test.tsx]: Why is this commented?

File: src/services/appCRUD/appConfService.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2020 Cognite AS
+import { MESSAGES } from 'constants/messages';
+import {
+  getMachineConfig,
+  saveMachineConfig,
+} from './appConfRepo/appConfFSRepo';
```
Comment (man-csl) [inline src/services/appCRUD/appConfService.ts:6]: Isn't there a better way to implement polymorphism in JS? Currently, service class is coupled with actual implementation.   Ideally it has to be decoupled and storage API must be injected to this some how.

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx (hunk 5 above)
Comment (rumeshwick) [reply src/components/MachineSelector/Configurator/MachineConfigurator.tsx:215]: is this for html(JSX) code?

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx (hunk 1 above)
Comment (rumeshwick) [reply src/components/MachineSelector/Configurator/MachineConfigurator.tsx]: This was actually used in a previous implementation and its no longer required to have a timeout now. Removed it

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx (hunk 2 above)
Comment (rumeshwick) [reply src/components/MachineSelector/Configurator/MachineConfigurator.tsx]: Updated both method name and comments. This is actually used to validate the saving asset ids list object.

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx (hunk 3 above)
Comment (rumeshwick) [reply src/components/MachineSelector/Configurator/MachineConfigurator.tsx:162]: Once we updated the asset external ids, state should be updated with the new asset(machine) list. I've added comments

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx (hunk 4 above)
Comment (rumeshwick) [reply src/components/MachineSelector/Configurator/MachineConfigurator.tsx]: Done

File: src/components/MachineSelector/Configurator/MachineConfigurator.tsx (hunk 5 above)
Comment (rumeshwick) [reply src/components/MachineSelector/Configurator/MachineConfigurator.tsx:215]: Done

File: src/components/widgets/timeSeries/TSFancyNumeric/TSFancyNumeric.test.tsx (hunk 6 above)
Comment (rumeshwick) [reply src/components/widgets/timeSeries/TSFancyNumeric/TSFancyNumeric.test.tsx]: I did a commit to test that travis build issue. I've reverted that. i'm not sure why still its shows as commented here. But in the file, it's not commented.

File: src/services/appCRUD/appConfService.ts (hunk 7 above)
Comment (rumeshwick) [reply src/services/appCRUD/appConfService.ts:6]: I followed the same way we save the widget configurations and there also this issue has been raised. Asanka has raised a task (#96 ) for that and we'll cover both scenarios under that task

**Review by man-csl — APPROVED** [review]

**Pull request merged**